### PR TITLE
Extend `eigenvalues`, `eigenvalues_with_multiplicity` to integer matrices

### PR DESCRIPTION
**Pull request by JohnAAbbott**
Also added some new tests for the extended functionality.

Resolves https://github.com/oscar-system/Oscar.jl/issues/4330.

### CODE REVIEW

File: test/matrix-test.jl [hunk 1]
```diff
@@ -59,3 +59,34 @@ end
   @test is_zero(reduce_mod(A, B))
   @test is_zero(reduce_mod(A, 2*B))
 end
+
+@testset "Eigenvalues of ZZMatrix" begin
+  A4 = matrix(ZZ,4,4, [6,-8,9,5,  5,-6,5,0,  0,6,5,4,  -6,8,-1,7]);
+  E4 = eigenvalues(A4);
+  @test length(E4) == 2;
+  @test E4 == [2,4] || E4 == [4,2]
```
Comment (lgoettgens) [inline test/matrix-test.jl:67]:
```suggestion
  @test issetequal(E4, [2,4])
```
a bit nicer to read

File: test/matrix-test.jl [hunk 2]
```diff
@@ -59,3 +59,34 @@ end
   @test is_zero(reduce_mod(A, B))
   @test is_zero(reduce_mod(A, 2*B))
 end
+
+@testset "Eigenvalues of ZZMatrix" begin
+  A4 = matrix(ZZ,4,4, [6,-8,9,5,  5,-6,5,0,  0,6,5,4,  -6,8,-1,7]);
```
Comment (lgoettgens) [inline test/matrix-test.jl:64]: please remove the trailing `;` from all of the added lines here

File: test/matrix-test.jl [hunk 3]
```diff
@@ -59,3 +59,34 @@ end
...
+  E4 = eigenvalues(A4);
+  @test length(E4) == 2;
+  @test E4 == [2,4] || E4 == [4,2]
+  E4m = eigenvalues_with_multiplicities(A4);
+  @test length(E4m) == 2;
+  @test E4m[1][2] == 1 || E4m[2][2] == 1
```
Comment (lgoettgens) [inline test/matrix-test.jl:70]:
```suggestion
  @test issetequal(E4m, [(2, 1), (4, 1)])
```
if I looked right at the matrix, this is what we expect from the tests here, right? If so, I think this change would make the test easier to read

File: test/matrix-test.jl [hunk 4]
```diff
@@ -59,3 +59,34 @@ end
...
+
+
+  A5 = matrix(ZZ,5,5, [0,0,1,-1,-1,  -1,-1,1,1,-1,  0,0,1,-1,-1,  -1,1,1,-1,-1,  1,1,-1,-1,1])
+  E5 = eigenvalues(A5);
+  @test length(E5) == 3;
+  @test sum(E5) == -1 && sum((e -> e^2), E5) == 5
```
Comment (lgoettgens) [inline test/matrix-test.jl:76]:
```suggestion
  @test sum(E5) == -1
  @test sum((e -> e^2), E5) == 5
```
splitting this into two tests makes it a bit easier to debug

File: test/matrix-test.jl [hunk 5]
```diff
@@ -59,3 +59,34 @@ end
...
+
+  E5m = eigenvalues_with_multiplicities(A5);
+  @test length(E5m) == 3;
+  @test E5m[1][2] + E5m[2][2] + E5m[3][2] == 5;
+
+  A6 = matrix(ZZ,6,6, [0,0,-1,0,1,0,  0,0,0,1,-1,0,  -1,0,0,0,1,0,  0,1,1,1,-1,0,  1,-1,1,1,-1,0,  1,-1,0,0,-1,-1]);
+  E6 = eigenvalues(A6);
+  @test length(E6) == 2; # -1 and +1
+  @test sum(E6) == 0 && sum((e -> e^2), E6) == 2
```
Comment (lgoettgens) [inline test/matrix-test.jl:85]:
```suggestion
  @test sum(E6) == 0
  @test sum((e -> e^2), E6) == 2
```

File: test/matrix-test.jl [hunk 6]
```diff
@@ -59,3 +59,34 @@ end
...
+
+  E6m = eigenvalues_with_multiplicities(A6);
+  @test length(E6m) == 2;
+  @test E6m[1][2] == 2 && E6m[2][2] == 2;
```
Comment (lgoettgens) [inline test/matrix-test.jl:89]:
```suggestion
  @test E6m[1][2] == 2
  @test E6m[2][2] == 2
```

**Review by lgoettgens — COMMENTED** [review]
Thanks @JohnAAbbott, just some minor style comments about the tests

File: test/matrix-test.jl (hunk 1 above)
Comment (JohnAAbbott) [reply test/matrix-test.jl]: OK, good suggestion

File: test/matrix-test.jl (hunk 3 above)
Comment (lgoettgens) [reply test/matrix-test.jl:70]: even without changing this to a `issetequal`, I think you wanna test that both multiplicities are 1 here, don't you?
```suggestion
  @test E4m[1][2] == 1
  @test E4m[2][2] == 1
```

Comment (JohnAAbbott) [reply test/matrix-test.jl]: Thanks!  You're right: I did mean `&&`.  Anyway I have just replaced it with `issetequal` which I think is easier to read and comprehend -- the reader knows immediately what the expected answer is.  Already pushed.  I was racing to be ready for an appointment at 16:00, so made a couple of mistakes (sigh).

**Review by fingolfin — APPROVED** [review]
Thanks!

**Pull request merged**